lib/parser.js

```diff
@@ -362,22 +366,25 @@ module.exports = class Parser {
     if (node === undefined) {
       node = this._t.peek();
     }
-    return this._posStack && node.location ? node.location.pos : -1;
+    return this._posStack && node.location ? node.location.start : -1;
   }
 
   getEnd(node) {
     return this._posStack && node.location ? node.location.end : -1;
   }
 
-  finish(node, pos, end) {
-    if (pos === undefined) {
-      pos = this.popPos();
-    }
-    if (end === undefined) {
-      end = this.getPos();
-    }
+  finish(node, start, end) {
     if (this._posStack) {
-      node.location = { pos, end };
+      if (start === undefined) {
+        start = this.popPos();
+      }
+      if (end === undefined) {
+        end =
+          this._t.previous === undefined
+            ? { line: 1, column: 0, offset: 0 }
+            : this._t.previous.location.end;
```

Comment:
From a memory perspective it makes sense to reuse the value of `end`, but you might want to be wary of anyone attempting to use the API and performing mutations on the AST. If someone were to attempt to insert a node and adjust the offsets of its preceding/trailing siblings they would be adjusting the end of more nodes than they might intend.

I'd recommend one of the following approaches:
- Clone the end each time (bad for memory)
- Mark the offsets as `readonly` in the declaration file (a soft check that only works for TS consumers).
- Mark the offsets as `readonly` in the declaration file and freeze the location (possibly overkill).

Reply:
Nice catch. I'll just clone it; I'm not that concerned about memory usage.